train.py

```diff
@@ -0,0 +1,215 @@
+import io
+import os
+
+import hydra
+import torch
+import torch.nn as nn
+from PIL import Image
+from tqdm import tqdm
+import torch.optim as optim
+import matplotlib.pyplot as plt
+from torch.utils.data import DataLoader
+from omegaconf import OmegaConf, DictConfig
+
+import wandb
+from utils.dataloaders import set_seed, create_dataloader
+from model.dcgan import Generator, Discriminator, weights_init
+
+
+def create_folders(cfg: DictConfig):
+    logger_dict = cfg.logger
+
+    for _, value in logger_dict.items():
+        if value[-1] == "/":
+            value = value[:-1]
+            os.makedirs(value, exist_ok=True)
+
+
+@torch.no_grad()
+def visualize_training(generator: Generator, fixed_noise: torch.Tensor, epoch: int, batch_idx: int, chart_path: str = "tmp/"):
+    generator.eval()
+
+    # Generate fake data using the generator and fixed noise
+    fake_data = generator(fixed_noise).detach().cpu().numpy()
+
+    # Create a figure and a grid of subplots
+    fig, axarr = plt.subplots(8, 2, figsize=(8, 16))
+
+    # Loop through each subplot to plot the 2-channel data
+    for i in range(8):
+        for j in range(2):
+            axarr[i, j].plot(fake_data[i, j, :])
+            axarr[i, j].set_title(f"Noise {i+1}, Channel {j+1}")
+            axarr[i, j].axis("off")
+            axarr[i, j].set_xticks([])
+            axarr[i, j].set_yticks([])
+
+    fig.suptitle(f"Epoch {epoch}, Batch {batch_idx}")
+    fig.tight_layout()
+    # save the figure to img/ folder
+    fig.savefig(f"{chart_path}epoch_{epoch}_batch_{batch_idx}.png")
+    return fig
+
+
+def save_progress_gif(fig_list, chart_path: str = "tmp/"):
+    imgs = []
+    for fig in fig_list:
+        buf = io.BytesIO()
+        fig.savefig(buf, format="png")
+        buf.seek(0)
+        img = Image.open(buf)
+        imgs.append(img)
+
+    # Save as an animated GIF
+    imgs[0].save(f"{chart_path}progress.gif", save_all=True, append_images=imgs[1:], loop=0, duration=500)
+
+
+def train_step(
+    generator: Generator,
+    discriminator: Discriminator,
+    train_loader: DataLoader,
+    gen_optimizer: torch.optim,
+    disc_optimizer: torch.optim,
+    criterion: nn.Module,
+    cfg: DictConfig,
+    fixed_noise: torch.Tensor,
+    epoch: int,
+) -> None:
+    generator.train()
+    discriminator.train()
+
+    real_label = 1.0
+    fake_label = 0.0
+
+    fig_list = []
+
+    progress_bar = tqdm(enumerate(train_loader), total=len(train_loader))
+    for batch_idx, batch in progress_bar:
+        real_data = batch["signal"].to(cfg.system.device)
+        batch_size = real_data.size(0)
+
+        # train discriminator
+        disc_optimizer.zero_grad()
+        label = torch.full((batch_size,), real_label, dtype=torch.float, device=cfg.system.device)
+        output = discriminator(real_data).view(-1)
+        discriminator_error_real = criterion(output, label)
+        discriminator_error_real.backward()
+        D_x = output.mean().item()
```

Comment:
```suggestion
        d_output_real_data = discriminator(real_data).view(-1)
        discriminator_error_real = criterion(d_output_real_data, label)
        discriminator_error_real.backward()
        D_x = d_output_real_data.mean().item()
```

`output` variable name seems overused in this loop, can you make it more descriptive for all the outputs processed here?

Also, is `D_x` a standard notation for GANs? Can you add to README some useful links to papers/tutorials/blogs for reference?

Reply:
I don't think it's standardized, I'll add comments for clarification and some references in README